contracts/SpokePoolV3Periphery.sol

```diff
@@ -289,323 +251,405 @@ contract SpokePoolV3Periphery is Lockable, MultiCaller {
+        uint32 exclusivityParameter,
+        bytes calldata message
     ) private {
-        _acrossInputToken.forceApprove(address(spokePool), _acrossInputAmount);
+        IERC20(inputToken).forceApprove(address(spokePool), inputAmount);
         spokePool.depositV3(
-            depositData.depositor,
-            depositData.recipient,
-            address(_acrossInputToken), // input token
-            depositData.outputToken, // output token
-            _acrossInputAmount, // input amount.
-            depositData.outputAmount, // output amount
-            depositData.destinationChainId,
-            depositData.exclusiveRelayer,
-            depositData.quoteTimestamp,
-            depositData.fillDeadline,
-            depositData.exclusivityParameter,
-            depositData.message
+            depositor,
+            recipient,
+            inputToken, // input token
+            outputToken, // output token
+            inputAmount, // input amount.
+            outputAmount, // output amount
+            destinationChainId,
+            exclusiveRelayer,
+            quoteTimestamp,
+            fillDeadline,
+            exclusivityParameter,
+            message
         );
     }
 
-    // This contract supports two variants of swap and bridge, one that allows one token and another that allows the caller to pass them in.
-    function _swapAndBridge(
-        address exchange,
-        bytes calldata routerCalldata,
-        uint256 swapTokenAmount,
-        uint256 minExpectedInputTokenAmount,
-        DepositData calldata depositData,
-        IERC20 _swapToken,
-        IERC20 _acrossInputToken
-    ) private {
+    /**
+     * @notice Swaps a token on the origin chain before depositing into the Across spoke pool atomically.
+     * @param swapAndDepositData The parameters to use when calling both the swap on an exchange and bridging via an Across spoke pool.
+     */
+    function _swapAndBridge(PeripherySigningLib.SwapAndDepositData calldata swapAndDepositData) private {
+        // Load variables we use multiple times onto the stack.
+        IERC20 _swapToken = IERC20(swapAndDepositData.swapToken);
+        IERC20 _acrossInputToken = IERC20(swapAndDepositData.depositData.inputToken);
+        PeripherySigningLib.TransferType _transferType = swapAndDepositData.transferType;
+        address _exchange = swapAndDepositData.exchange;
+        uint256 _swapTokenAmount = swapAndDepositData.swapTokenAmount;
+
         // Swap and run safety checks.
         uint256 srcBalanceBefore = _swapToken.balanceOf(address(this));
         uint256 dstBalanceBefore = _acrossInputToken.balanceOf(address(this));
 
-        _swapToken.forceApprove(exchange, swapTokenAmount);
+        // The exchange will either receive funds from this contract via a direct transfer, an approval to spend funds on this contract, or via an
+        // EIP1271 permit2 signature.
+        if (_transferType == PeripherySigningLib.TransferType.Approval)
```

Comment:
Can you add unit tests for these three `TransferType` flows? I also think we should move `TransferType` into this contract as its not really a "signing" related struct